Context Menu shows Exit Full Screen instead of Exit Viewer

danae404 opened the pull request:
#### de1667dea3dd915fbd6610e7ffc94a7b23a3b29c
<pre>
Context Menu shows Exit Full Screen instead of Exit Viewer
<a href="https://bugs.webkit.org/show_bug.cgi?id=275460">https://bugs.webkit.org/show_bug.cgi?id=275460</a>
<a href="https://rdar.apple.com/126300924">rdar://126300924</a>

Reviewed by Jer Noble.

This patch adds a context menu item to video elements
that enters and exits Viewer mode. This patch also
removes the toggle fullscreen context menu item when
the video is in Viewer mode.

* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebCore/page/ContextMenuController.cpp:
(WebCore::ContextMenuController::contextMenuItemSelected):
(WebCore::ContextMenuController::populate):
(WebCore::ContextMenuController::checkOrEnableIfNeeded const):
* Source/WebCore/platform/ContextMenuItem.cpp:
(WebCore::isValidContextMenuAction):
* Source/WebCore/platform/ContextMenuItem.h:
* Source/WebCore/platform/LocalizedStrings.h:
* Source/WebCore/platform/cocoa/LocalizedStringsCocoa.mm:
(WebCore::contextMenuItemTagEnterVideoViewer):
(WebCore::contextMenuItemTagExitVideoViewer):
* Source/WebCore/rendering/HitTestResult.cpp:
(WebCore::HitTestResult::mediaIsInVideoViewer const):
(WebCore::HitTestResult::toggleVideoViewer const):
* Source/WebCore/rendering/HitTestResult.h:
* Source/WebKit/Shared/API/c/WKContextMenuItemTypes.h:
* Source/WebKit/Shared/API/c/WKSharedAPICast.h:
(WebKit::toAPI):
(WebKit::toImpl):
* Source/WebKit/UIProcess/API/Cocoa/WKMenuItemIdentifiers.mm:
* Source/WebKit/UIProcess/API/Cocoa/WKMenuItemIdentifiersPrivate.h:
* Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.mm:
(WebKit::menuItemIdentifier):
* Source/WebKitLegacy/mac/WebView/WebHTMLView.mm:
(toAction):
(toTag):
* Source/WebKitLegacy/mac/WebView/WebUIDelegatePrivate.h:

Canonical link: <a href="https://commits.webkit.org/280128@main">https://commits.webkit.org/280128@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/61d3bc9cd6d19915562d0b422342d6b10101c55d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/55763 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/35086 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/8230 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/58747 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/6194 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/57889 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/42708 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/6392 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/44906 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4263 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/57792 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/32984 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/48066 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26040 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29770 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/5395 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/4337 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/51741 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/5664 "Found 1 new test failure: imported/w3c/web-platform-tests/resource-timing/status-codes-create-entry.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/60338 "Built successfully") | 
| | [⏳ 🛠 vision ](https://ews-build.webkit.org/#/builders/visionOS-1-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/5794 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/52335 "Passed tests") | 
| | [⏳ 🛠 vision-sim ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/48137 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/51832 "Passed tests") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8236 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/30917 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/32002 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/33083 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/31749 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/9b9488571c5c0f27c91be2e9f909c38a19ec6aee. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/29798

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/aba38a634b99c349e5a6ebf29377958183741754. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/29798

rr-codes commented on Source/WebCore/WebCore.xcodeproj/project.pbxproj:
```diff
@@ -3,7 +3,7 @@
 	classes = {
 	};
-	objectVersion = 54;
+	objectVersion = 55;
```
Is this ok?

rr-codes commented on Source/WebCore/en.lproj/Localizable.strings:
```diff
@@ -175,6 +175,9 @@
 /* Allow button title in speech recognition prompt */
 "Allow (speechrecognition)" = "Allow";
 
+/* Allow */
```
are these ok to add?

pxlcoder replied:
This isn't being added, just moved. The script sorts alphabetically.

Though we should preserve the existing comment for this string, which was `Display USDZ file`.

rr-codes commented on Source/WebCore/rendering/HitTestResult.cpp:
```diff
@@ -542,6 +542,32 @@ void HitTestResult::enterFullscreenForVideo() const
+void HitTestResult::toggleVideoViewer() const
+{
+#if PLATFORM(MAC) && ENABLE(VIDEO) && ENABLE(VIDEO_PRESENTATION_MODE)
+    auto* mediaElement(this->mediaElement());
+    auto* videoElement = dynamicDowncast<HTMLVideoElement>(*mediaElement);
```
I think these should be RefPtr's maybe?

danae404 replied:
I see a lot of examples in this file of raw pointers being used (for example, HitTestResult::toggleEnhancedFullscreenForVideo()). I am not sure if this is for a particular reason, or if it would still be best practice to use RefPtr's. @pxlcoder do you have thoughts?

pxlcoder replied:
All new code should use `RefPtr`s. We are slowly working towards adopting smart pointers everywhere (which is why others in this file have not yet been updated).

rr-codes commented on Source/WebCore/rendering/HitTestResult.cpp:
```diff
@@ -542,6 +542,32 @@ void HitTestResult::enterFullscreenForVideo() const
+        return;
+
+    UserGestureIndicator indicator(IsProcessingUserGesture::Yes, &mediaElement->document());
+    if (videoElement->webkitPresentationMode() == HTMLVideoElement::VideoPresentationMode::InWindow)
+        videoElement->webkitSetPresentationMode(HTMLVideoElement::VideoPresentationMode::Inline);
+    else
+        videoElement->webkitSetPresentationMode(HTMLVideoElement::VideoPresentationMode::InWindow);
```
```suggestion
    auto newMode = videoElement->webkitPresentationMode() == HTMLVideoElement::VideoPresentationMode::InWindow ? HTMLVideoElement::VideoPresentationMode::Inline : HTMLVideoElement::VideoPresentationMode::InWindow;
    
    videoElement->webkitSetPresentationMode(newMode);
```

rr-codes commented on Source/WebKit/UIProcess/API/Cocoa/WKMenuItemIdentifiersPrivate.h:
```diff
@@ -51,6 +51,7 @@ WK_EXTERN NSString * const _WKMenuItemIdentifierSearchWeb WK_API_AVAILABLE(macos
 WK_EXTERN NSString * const _WKMenuItemIdentifierShowHideMediaControls WK_API_AVAILABLE(macos(10.12), ios(10.0));
 WK_EXTERN NSString * const _WKMenuItemIdentifierShowHideMediaStats WK_API_AVAILABLE(macos(13.3), ios(16.4));
 WK_EXTERN NSString * const _WKMenuItemIdentifierToggleEnhancedFullScreen WK_API_AVAILABLE(macos(10.14), ios(12.0));
+WK_EXTERN NSString * const _WKMenuItemIdentifierToggleVideoViewer WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA));
```
I wonder if these can be the actual versions now

danae404 replied:
@pxlcoder do you know the answer to this?

pxlcoder replied:
I think this is fine as-is. They will already resolve to the correct actual versions right now due to changes made in rdar://127507228.

And we can update all TBAs for the new release together.

pxlcoder commented on Source/WebCore/en.lproj/Localizable.strings:
```diff
@@ -1096,6 +1102,9 @@
 /* Title for Open in External Application Link action button */
 "Open in “%@”" = "Open in “%@”";
 
+/* Open this 3D model? */
```
We should probably preserve the original comment, which was `Open 3D object in a new window?`.

pxlcoder commented on Source/WebCore/en.lproj/Localizable.strings:
```diff
@@ -286,12 +289,12 @@
 /* Button title in Device Orientation Permission API prompt */
 "Cancel (device motion and orientation access)" = "Cancel";
 
+/* Cancel */
```
Same comment about preserving the original comment. `Cancel USDZ file`.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/590540ce1d0904271e9bb23aeb5cb7f8a1350c09. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/29798

rr-codes commented on Source/WebCore/rendering/HitTestResult.cpp:
```diff
@@ -542,6 +542,31 @@ void HitTestResult::enterFullscreenForVideo() const
+{
+#if PLATFORM(MAC) && ENABLE(VIDEO) && ENABLE(VIDEO_PRESENTATION_MODE)
+    RefPtr mediaElement(this->mediaElement());
+    RefPtr videoElement = dynamicDowncast<HTMLVideoElement>(*mediaElement);
```
I think this is unsafe is `mediaElement` is ever null?

danae404 replied:
I think dynamicDowncast returns a nullptr if mediaElement is null

jernoble replied:
I think you need to do:
```
RefPtr videoElement = dynamicDowncast<HTMLVideoElement>(mediaElement);
```
I.e., without the `*` dereference.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/5ab3302901170f40578d19dc4f346c1183ca56e1. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/29798

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/61d3bc9cd6d19915562d0b422342d6b10101c55d. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/29798

jernoble commented on Source/WebCore/rendering/HitTestResult.cpp:
```diff
@@ -542,6 +542,31 @@ void HitTestResult::enterFullscreenForVideo() const
+void HitTestResult::toggleVideoViewer() const
+{
+#if PLATFORM(MAC) && ENABLE(VIDEO) && ENABLE(VIDEO_PRESENTATION_MODE)
+    RefPtr mediaElement(this->mediaElement());
+    RefPtr videoElement = dynamicDowncast<HTMLVideoElement>(mediaElement);
```
Nits: Early returns are always ok, so this could bail after the first `RefPtr`. Also, this could be written either as a simple assignment, or with the modern, curly-brace initializer:

```
RefPtr mediaElement { this->mediaElement() };
if (!mediaElement)
    return;

RefPtr videoElement = dynamicDowncast<HTMLVideoElement>(mediaElement);
```

But this _would_ null-check `mediaElement` twice, so you could add back in the dereference to `mediaElement` inside `dynamicDowncast()`. Just a suggestion though.

danae404 committed:
Context Menu shows Exit Full Screen instead of Exit Viewer
https://bugs.webkit.org/show_bug.cgi?id=275460
rdar://126300924

Reviewed by Jer Noble.

This patch adds a context menu item to video elements
that enters and exits Viewer mode. This patch also
removes the toggle fullscreen context menu item when
the video is in Viewer mode.

* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebCore/page/ContextMenuController.cpp:
(WebCore::ContextMenuController::contextMenuItemSelected):
(WebCore::ContextMenuController::populate):
(WebCore::ContextMenuController::checkOrEnableIfNeeded const):
* Source/WebCore/platform/ContextMenuItem.cpp:
(WebCore::isValidContextMenuAction):
* Source/WebCore/platform/ContextMenuItem.h:
* Source/WebCore/platform/LocalizedStrings.h:
* Source/WebCore/platform/cocoa/LocalizedStringsCocoa.mm:
(WebCore::contextMenuItemTagEnterVideoViewer):
(WebCore::contextMenuItemTagExitVideoViewer):
* Source/WebCore/rendering/HitTestResult.cpp:
(WebCore::HitTestResult::mediaIsInVideoViewer const):
(WebCore::HitTestResult::toggleVideoViewer const):
* Source/WebCore/rendering/HitTestResult.h:
* Source/WebKit/Shared/API/c/WKContextMenuItemTypes.h:
* Source/WebKit/Shared/API/c/WKSharedAPICast.h:
(WebKit::toAPI):
(WebKit::toImpl):
* Source/WebKit/UIProcess/API/Cocoa/WKMenuItemIdentifiers.mm:
* Source/WebKit/UIProcess/API/Cocoa/WKMenuItemIdentifiersPrivate.h:
* Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.mm:
(WebKit::menuItemIdentifier):
* Source/WebKitLegacy/mac/WebView/WebHTMLView.mm:
(toAction):
(toTag):
* Source/WebKitLegacy/mac/WebView/WebUIDelegatePrivate.h:

Canonical link: https://commits.webkit.org/280128@main
